Runtime Error when using span on LeetCode

## Steps to Reproduce

1. Open <https://leetcode.com/problems/longest-palindromic-substring/>.
2. Copy and Paste my following code.
3. Click <kbd>Run Code</kbd>

I don't think my code has an algorithm issue because changing span to string works.

```c#
public class Solution {
    public string LongestPalindrome(string s) // DP
    {
        ReadOnlySpan<char> ss = s.AsSpan();
        ReadOnlySpan<char> str = "";
        for (int i = 0; i < s.Length; i++)
        {
            var r = Expand(ss, i - 1, i + 1);
            if (r.Length > str.Length)
                str = r;
            r = Expand(ss, i, i + 1);
            if (r.Length > str.Length)
                str = r;
        }
        return str.ToString();
    }
    internal ReadOnlySpan<char> Expand(ReadOnlySpan<char> s, int from, int to)
    {
        while (from >= 0 && to <= s.Length - 1)
        {
            if (s[from] != s[to])
                break;
            from--;
            to++;
        }

        return from + 1 == to ? "" : s.Slice(from + 1, to - from - 1);
    }
}
```

### Current Behavior

Runtime Error

### Expected Behavior

Pass

## On which platforms did you notice this

[ ] macOS
[x] Linux
[ ] Windows

## Stacktrace

![图片](https://user-images.githubusercontent.com/24759802/54805094-d26fad00-4cb0-11e9-9730-d6c321b04126.png)

The stacktrace text on LeetCode seems can't be copied. If you want to see it, you must do it yourself.

---

In case this is a LeetCode issue, I tried my code in Ubuntu vm. But this is my first time using mono, I encountered many errors like "couldn't found System.Memory" or "Could not load file or assembly System.Runtime.Loader".

**So if you can compile and xunit test pass my following code, just close this issue.**

<details>
<summary>Full Code</summary>

```c#
using System;
using Xunit;

namespace LongestPalindromicSubstring
{
    public interface ISolution { string LongestPalindrome(string s); }
    class Solution : ISolution
    {
        public string LongestPalindrome(string s) // BF
        {
            if (s.Length == 1)
                return s;
            var result = string.Empty.AsSpan();
            var span = s.AsSpan();
            for (int i = 0; i < s.Length - 1; i++)
                for (int j = i + 1; j <= s.Length; j++)
                {
                    var ss = span.Slice(i, j - i);
                    if (IsPalindromicString(ss) == true && ss.Length > result.Length)
                        result = ss;
                }
            return result.ToString();
        }
        internal bool IsPalindromicString(ReadOnlySpan<char> s)
        {
            if (s.Length == 1)
                return true;
            for (int i = 0; i < s.Length / 2; i++)
                if (s[i] != s[s.Length - i - 1])
                    return false;
            return true;
        }
    }

    class Solution2 : ISolution
    {
        public string LongestPalindrome(string s) // DP
        {
            ReadOnlySpan<char> ss = s.AsSpan();
            ReadOnlySpan<char> str = "";
            for (int i = 0; i < s.Length; i++)
            {
                var r = Expand(ss, i - 1, i + 1);
                if (r.Length > str.Length)
                    str = r;
                r = Expand(ss, i, i + 1);
                if (r.Length > str.Length)
                    str = r;
            }
            return str.ToString();
        }
        internal ReadOnlySpan<char> Expand(ReadOnlySpan<char> s, int from, int to) // 后两者表示将要评估的位置
        {
            while (from >= 0 && to <= s.Length - 1)
            {
                if (s[from] != s[to])
                    break;
                from--;
                to++;
            }

            return from + 1 == to ? "" : s.Slice(from + 1, to - from - 1);
        }
    }

    abstract
    public class MultiTest
    {
        protected virtual ISolution GetSo => new Solution2();

        [Theory]
        [InlineData("babad", "bab"), InlineData("cbbd", "bb"), InlineData("cbbb", "bbb")]
        [InlineData("a", "a"), InlineData("bb", "bb")]
        public void Test(string input, string expect)
        {
            var so = GetSo;
            string result = so.LongestPalindrome(input);
            Assert.Equal(expect, result);
        }
    }
    public class Test1 : MultiTest { protected override ISolution GetSo => new Solution(); }
    public class Test2 : MultiTest { protected override ISolution GetSo => new Solution2(); }

    public class BFTest
    {
        [Theory]
        [InlineData("bab", true), InlineData("bb", true)]
        [InlineData("baba", false)]
        [InlineData("a", true)]
        void IsPalindromicStringTest(string s, bool expect)
        {
            var so = new Solution();
            bool result = so.IsPalindromicString(s);
            Assert.Equal(expect, result);
        }
    }

    public class DPTest
    {
        [Theory]
        [InlineData("a", -1, 1), InlineData("bab", 0, 2),
        InlineData("bb", 0, 1), InlineData("abba", 1, 2)]
        void ExpandTest1(string s, int from, int to)
        {
            var so = new Solution2();
            string r = so.Expand(s, from, to).ToString();
            Assert.Equal(s, r);
        }
        [Fact]
        void ExpandTest2()
        {
            var so = new Solution2();
            string r = so.Expand("cb", 0, 1).ToString();
            Assert.Equal("", r);
        }
        [Fact]
        void ExpandTest3()
        {
            var so = new Solution2();
            string r = so.Expand("cbbb", 1, 2).ToString();
            Assert.Equal("bb", r);
        }
    }
}
```
</details>

But I would also appreciate it if you could tell me how to use mono run xunit.

<details>
<summary>Errors I encountered</summary>

```bash
nuget restore LeetCode.csproj
nuget install xunit.runner.console
msbuild /p:Configuration=Debug
mono ./xunit.runner.console.2.4.1/tools/netcoreapp2.0/xunit.console.dll ./bin/Debug/netcoreapp2.1/LeetCode.dll

System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.Loader, Version=4.0.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
File name: 'System.Runtime.Loader, Version=4.0.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
[ERROR] FATAL UNHANDLED EXCEPTION: System.IO.FileNotFoundException: Could not load file or assembly 'System.Runtime.Loader, Version=4.0.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' or one of its dependencies.
File name: 'System.Runtime.Loader, Version=4.0.2.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'

mono ./xunit.runner.console.2.4.1/tools/net472/xunit.console.exe ./bin/Debug/netcoreapp2.1/LeetCode.dll

xUnit.net Console Runner v2.4.1 (64-bit Desktop .NET 4.7.2, runtime: 4.0.30319.42000)
System.InvalidOperationException: Unknown test framework: could not find xunit.dll (v1) or xunit.execution.*.dll (v2) in /root/TEST/bin/Debug/netcoreapp2.1

msbuild /t:Publish
mono ./xunit.runner.console.2.4.1/tools/net472/xunit.console.exe ./bin/Debug/netcoreapp2.1/publish/LeetCode.dll

System.IO.FileNotFoundException : Could not load file or assembly 'System.Memory, Version=4.1.0.0, Culture=neutral, PublicKeyToken=cc7b13ffcd2ddd51' or one of its dependencies.
```
</details>

--- a/mono/arch/amd64/amd64-codegen.h
+++ b/mono/arch/amd64/amd64-codegen.h
@@ -450,15 +450,18 @@ typedef union {
 	} while (0)
 
 
-#define amd64_lea_membase_body(inst,reg,basereg,disp)	\
+#define amd64_lea_membase_body(inst,reg,basereg,disp,width)	\
 	do {	\
-		amd64_emit_rex(inst, 8, (reg), 0, (basereg)); \
+		amd64_emit_rex(inst, width, (reg), 0, (basereg)); \
 		*(inst)++ = (unsigned char)0x8d;	\
 		amd64_membase_emit ((inst), (reg), (basereg), (disp));	\
 	} while (0)
 
+#define amd64_lea4_membase(inst,reg,basereg,disp) \
+	amd64_lea_membase_body((inst), (reg), (basereg), (disp), 4)
+
 #define amd64_lea_membase(inst,reg,basereg,disp) \
-	amd64_lea_membase_body((inst), (reg), (basereg), (disp))
+	amd64_lea_membase_body((inst), (reg), (basereg), (disp), 8)
 
 /* Instruction are implicitly 64-bits so don't generate REX for just the size. */
 #define amd64_push_reg(inst,reg)	\

--- a/mono/mini/cpu-amd64.md
+++ b/mono/mini/cpu-amd64.md
@@ -336,6 +336,7 @@ x86_push_membase: src1:b len:8
 x86_push_obj: src1:b len:40
 x86_lea: dest:i src1:i src2:i len:8
 x86_lea_membase: dest:i src1:i len:11
+amd64_lea_membase: dest:i src1:i len:11
 x86_xchg: src1:i src2:i clob:x len:2
 x86_fpop: src1:f len:3
 x86_seteq_membase: src1:b len:9

--- a/mono/mini/mini-amd64.c
+++ b/mono/mini/mini-amd64.c
@@ -2460,7 +2460,7 @@ mono_arch_emit_outarg_vt (MonoCompile *cfg, MonoInst *ins, MonoInst *src)
 		mini_emit_memcpy (cfg, load->dreg, 0, src->dreg, 0, size, TARGET_SIZEOF_VOID_P);
 
 		if (ainfo->pair_storage [0] == ArgInIReg) {
-			MONO_INST_NEW (cfg, arg, OP_X86_LEA_MEMBASE);
+			MONO_INST_NEW (cfg, arg, OP_AMD64_LEA_MEMBASE);
 			arg->dreg = mono_alloc_ireg (cfg);
 			arg->sreg1 = load->dreg;
 			arg->inst_imm = 0;
@@ -3201,7 +3201,7 @@ mono_arch_peephole_pass_1 (MonoCompile *cfg, MonoBasicBlock *bb)
 				 * sreg1==dreg restriction. inst_imm > 0 is needed since LEA sign-extends 
 				 * its operand to 64 bit.
 				 */
-				ins->opcode = OP_X86_LEA_MEMBASE;
+				ins->opcode = ins->opcode == OP_IADD_IMM ? OP_X86_LEA_MEMBASE : OP_AMD64_LEA_MEMBASE;
 				ins->inst_basereg = ins->sreg1;
 			}
 			break;
@@ -5074,6 +5074,9 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
 			amd64_lea_memindex (code, ins->dreg, ins->sreg1, ins->inst_imm, ins->sreg2, ins->backend.shift_amount);
 			break;
 		case OP_X86_LEA_MEMBASE:
+			amd64_lea4_membase (code, ins->dreg, ins->sreg1, ins->inst_imm);
+			break;
+		case OP_AMD64_LEA_MEMBASE:
 			amd64_lea_membase (code, ins->dreg, ins->sreg1, ins->inst_imm);
 			break;
 		case OP_X86_XCHG:

--- a/mono/mini/mini-ops.h
+++ b/mono/mini/mini-ops.h
@@ -1190,6 +1190,7 @@ MINI_OP(OP_AMD64_AND_MEMBASE_IMM,        "amd64_and_membase_imm", NONE, IREG, NO
 MINI_OP(OP_AMD64_OR_MEMBASE_IMM,         "amd64_or_membase_imm", NONE, IREG, NONE)
 MINI_OP(OP_AMD64_XOR_MEMBASE_IMM,        "amd64_xor_membase_imm", NONE, IREG, NONE)
 MINI_OP(OP_AMD64_MUL_MEMBASE_IMM,        "amd64_mul_membase_imm", NONE, IREG, NONE)
+MINI_OP(OP_AMD64_LEA_MEMBASE,            "amd64_lea_membase", IREG, IREG, NONE)
 
 MINI_OP(OP_AMD64_ADD_REG_MEMBASE,        "amd64_add_reg_membase", IREG, IREG, IREG)
 MINI_OP(OP_AMD64_SUB_REG_MEMBASE,        "amd64_sub_reg_membase", IREG, IREG, IREG)

